qlib/contrib/model/pytorch_hgats_ts.py

```diff
@@ -0,0 +1,529 @@
+                    )
+                ]  # [~#stocks, #industries]
+
+            with torch.no_grad():
+                pred = self.GAT_model(feature.float(), GH).detach().cpu().numpy()
+
+            preds.append(pred)
+
+        return pd.Series(
+            np.concatenate(preds),
+            index=dl_test.get_index()[: len(np.concatenate(preds))],
+        )
+
+
+class GATModel(nn.Module):
+    def __init__(
+        self,
+        d_feat=6,
+        hidden_size=64,
+        num_layers=2,
+        dropout=0.0,
+        num_ind=29,  # number of industries
+        base_model="GRU",
+    ):
+        super().__init__()
+
+        if base_model == "GRU":
+            self.rnn = nn.GRU(
+                input_size=d_feat,
+                hidden_size=hidden_size,
+                num_layers=num_layers,
+                batch_first=True,
+                dropout=dropout,
+            )
+        elif base_model == "LSTM":
+            self.rnn = nn.LSTM(
+                input_size=d_feat,
+                hidden_size=hidden_size,
+                num_layers=num_layers,
+                batch_first=True,
+                dropout=dropout,
+            )
+        else:
+            raise ValueError("unknown base model name `%s`" % base_model)
+
+        self.hidden_size = hidden_size
+        self.num_ind = num_ind
+        self.d_feat = d_feat
+        self.transformationi = nn.Linear(self.hidden_size, self.hidden_size)
+        self.transformationj = nn.Linear(self.num_ind, self.num_ind)
+        self.a = nn.Parameter(torch.randn(self.hidden_size + self.num_ind, 1))
+        self.a.requires_grad = True
+        self.fc = nn.Linear(self.hidden_size, self.hidden_size)
+        self.fc_out = nn.Linear(hidden_size, 1)
+        self.leaky_relu = nn.LeakyReLU()
+        self.softmax = nn.Softmax(dim=1)
+        self.device = torch.device("cuda:0")
+
+    def cal_attention(self, x, y):
+        x = self.transformationi(x)  # x <- hidden
+        y = self.transformationj(y)  # y <- GH
+
+        sample_num = x.shape[0]  # num of securities
+        dim1 = x.shape[1]  # num of temporal features
+        dim2 = y.shape[1]  # num of industries
+        e_x = x.expand(dim2, sample_num, dim1)
+        e_y = y.expand(dim2, sample_num, dim2)
+        attention_in = torch.cat((e_x, e_y), 2).view(-1, dim1 + dim2)  # P i || P j
+        self.a_t = torch.t(self.a)
+        attention_out = self.a_t.mm(torch.t(attention_in)).view(dim2, sample_num)
+        attention_out = self.leaky_relu(attention_out)
+        att_weight = self.softmax(attention_out)
+        return torch.t(att_weight)
+
+    def forward(self, x, GH):
+
+        out, _ = self.rnn(x)
+        hidden = out[:, -1, :]  # [~#stocks, #features]
+        att_weight = self.cal_attention(
+            hidden, torch.Tensor(GH).to(self.device)
+        )  # [~#stocks, #industries]
+
+        De = torch.Tensor(pinv(np.diag(GH.sum(axis=0)))).to(self.device)
```

Comment:
The device should be managed in `class GATs(Model):`

Reply:
done